pytorch_lightning/plugins/training_type/ipu.py

```diff
@@ -0,0 +1,239 @@
+        return obj
+
+    @property
+    def lightning_module(self) -> Optional[LightningModule]:
+        model = self.model.module if isinstance(self.model, poptorch.PoplarExecutor) else self.model
+        return model.module if isinstance(model, LightningIPUModule) else model
+
+    def pre_dispatch(self) -> None:
+        '''
+        The issue here is we assume we're training.
+        What if we're not training?
+        I say
+        '''
+        if self.half:
+            log.info('Using full 16bit precision, converting LightningModule weights to FP16.')
+            self.model = self.model.half()
+        precision = self.lightning_module.trainer.accelerator.precision_plugin.precision
+        precision = 16 if self.half else precision
+
+        model = LightningIPUModule(self.lightning_module, precision)
+        self.model = model
+
+        # Separate models are instantiated for different stages, but they share the same weights on host.
+        # When validation/test models are run, they sync weights first.
+
+        self.poptorch_wrapped_models = {}
+        if self.lightning_module.trainer.training:
+            # Create model for training which will run training.
+            optimizer = self.lightning_module.trainer.optimizers[0]
+            self.poptorch_wrapped_models['train'] = poptorch.trainingModel(
+                model=model, options=self._create_opts(training=True), optimizer=optimizer
+            )
+        for x in ('val', 'test', 'predict'):
+            self.poptorch_wrapped_models[x] = poptorch.inferenceModel(
+                model=model,
+                options=self._create_opts(training=False),
+            )
+
+    @property
+    def replication_factor(self):
+        return len(self.parallel_devices)
+
+    def _create_opts(self, training):
+        opts = poptorch.Options()
+        opts.deviceIterations(self.device_iterations)
+        opts.replicationFactor(self.replication_factor)
+        gradient_accumulation = self.lightning_module.trainer.accumulate_grad_batches if training else 1
+        opts.Training.gradientAccumulation(gradient_accumulation)
+        opts.autoRoundNumIPUs(self.autoround_num_ipus)
+
+        # todo (sean): unsure if this is necessary but to be safe.
+        if os.environ.get("PL_GLOBAL_SEED"):
+            opts.randomSeed(int(os.environ["PL_GLOBAL_SEED"]))
+        return opts
+
+    def on_reset_train_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
+
+    def on_reset_val_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
+
+    def on_reset_test_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
+
+    def on_reset_predict_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
```

Comment:
Need to be pulled into the base plugin in a separate PR if we're comfortable with adding these hooks. Basically the reason these have been introduced is `process_dataloader` assumes that the dataset size doesn't change. If the size does change, the progress bar size is messed up.

These hooks happen early enough in the code that progress bar is correct. I also looked into moving `process_dataloader` but this is finnicky.

Reply:
instead of the multiple hooks, the method `process_dataloader` could simply be the one that goes into the base class.

Reply:
There was some confusion here; `process_dataloader` already exists in the `TrainingTypePlugin`